src/io/io_session.f90

```diff
@@ -0,0 +1,215 @@
+    class(io_session_t), intent(inout) :: self
+    character(len=*), intent(in) :: variable_name
+    integer, intent(in) :: value
+
+    if (.not. self%is_open) error stop "IO session not open"
+    if (.not. self%is_write_mode) error stop "IO session not in write mode"
+    call self%writer%write_data(variable_name, value, self%file)
+  end subroutine write_data_integer
+
+  subroutine write_data_real(self, variable_name, value)
+    class(io_session_t), intent(inout) :: self
+    character(len=*), intent(in) :: variable_name
+    real(dp), intent(in) :: value
+
+    if (.not. self%is_open) error stop "IO session not open"
+    if (.not. self%is_write_mode) error stop "IO session not in write mode"
+    call self%writer%write_data(variable_name, value, self%file)
+  end subroutine write_data_real
+
+  subroutine write_data_array_3d( &
+    self, variable_name, array, start_dims, count_dims &
+    )
+    class(io_session_t), intent(inout) :: self
+    character(len=*), intent(in) :: variable_name
+    real(dp), intent(in) :: array(:, :, :)
+    integer(i8), intent(in), optional :: start_dims(3)
+    integer(i8), intent(in), optional :: count_dims(3)
+
+    if (.not. self%is_open) error stop "IO session not open"
+    if (.not. self%is_write_mode) error stop "IO session not in write mode"
+    call self%writer%write_data( &
+      variable_name, array, self%file, &
+      start_dims=start_dims, count_dims=count_dims &
+      )
+  end subroutine write_data_array_3d
+
+  subroutine session_write_attribute(self, attribute_name, attribute_value)
+    class(io_session_t), intent(inout) :: self
+    character(len=*), intent(in) :: attribute_name
+    character(len=*), intent(in) :: attribute_value
+
+    if (.not. self%is_open) error stop "IO session not open"
+    if (.not. self%is_write_mode) error stop "IO session not in write mode"
+    call self%writer%write_attribute( &
+      attribute_name, attribute_value, self%file &
+      )
+  end subroutine session_write_attribute
+
+  subroutine session_open(self, filename, comm, mode)
+    class(io_session_t), intent(inout) :: self
+    character(len=*), intent(in) :: filename
+    integer, intent(in) :: comm
+    integer, intent(in), optional :: mode
+
+    integer :: use_mode
+
+    if (self%is_open) error stop "IO session already open"
+
+    ! Default to read mode if not specified
+    use_mode = io_mode_read
+    if (present(mode)) use_mode = mode
+
+    if (use_mode == io_mode_read) then
+      call allocate_io_reader(self%reader)
+      call self%reader%init(comm, "session_reader")
+      self%file => self%reader%open(filename, io_mode_read, comm)
+      self%is_write_mode = .false.
+    else if (use_mode == io_mode_write) then
+      call allocate_io_writer(self%writer)
+      call self%writer%init(comm, "session_writer")
+      self%file => self%writer%open(filename, io_mode_write, comm)
+      self%is_write_mode = .true.
+    else
+      error stop "Invalid I/O mode"
+    end if
```

Comment:
wrt above comment the user could try to use read operations from a session in write mode and vice-versa but this would not be caught until runtime.